package/src/prmon.cpp

```diff
@@ -67,31 +65,30 @@ int MemoryMonitor(const pid_t mpid, const std::string filename,
   signal(SIGCHLD, SignalChildHandler);
 
   // This is the vector of all monitoring components
-  std::vector<Imonitor*> monitors{};
-
-  // Number of processes and threads monitoring
-  countmon count_monitor{};
-  monitors.push_back(&count_monitor);
-
-  // Wall clock monitoring
-  wallmon wall_monitor{};
-  monitors.push_back(&wall_monitor);
-
-  // CPU monitoring
-  cpumon cpu_monitor{};
-  monitors.push_back(&cpu_monitor);
-
-  // Memory monitoring
-  memmon mem_monitor{};
-  monitors.push_back(&mem_monitor);
-
-  // IO monitoring
-  iomon io_monitor{};
-  monitors.push_back(&io_monitor);
-
-  // Network monitoring
-  netmon network_monitor{netdevs};
-  monitors.push_back(&network_monitor);
+  std::vector<std::unique_ptr<Imonitor>> monitors{};
+
+  auto registered_monitors = registry::Registry<Imonitor>::list_registered();
+  wallmon* wallclock_monitor_p = nullptr; // Special: we need to use the wallclock monitor
+                                          // inside the main prmon loop, so we will grab a
+                                          // pointer to it
+  for (const auto& class_name : registered_monitors) {
+    std::unique_ptr<Imonitor> new_monitor_p(registry::Registry<Imonitor>::create(class_name));
+    if (new_monitor_p) {
+      if (class_name == "wallmon") {
+        // The monitors' scope is the same as this pointer, and it is only used
+        // inside the main monitoring loop, so no need to worry about using
+        // this after the object is destroyed
+        wallclock_monitor_p = static_cast<wallmon*>(new_monitor_p.get());
+      }
+      monitors.push_back(std::move(new_monitor_p));
+    } else {
+      std::cerr << "Registration of monitor " << class_name << " FAILED" << std::endl;
+    }
+  }
+  if (wallclock_monitor_p == nullptr) {
+    std::cerr << "Failed to initialise wallclock monitoring class" << std::endl;
+    exit(EXIT_FAILURE);
+  }
```

Comment:
I agree, this bit for the `wallmon` is not so nice. I wonder if we can implement a `get` method in the factory that'll return the singleton based on the `class_name`. That way, perhaps these bits might be tied up a bit.

Reply:
I don't favour the idea that the registry itself keeps track of the instantiated classes - singletons are just a bit fragile, so I prefer the fact that the registry is basically compile time constructed. 

Here, keeping track of what prmon is using is the job of the `monitors` vector of unique pointers. Of course instead of a vector we could make that an `unordered_map` so that we could have the name of the component as a key. There's a penalty in iterating over a map instead of a vector - maybe irrelevant though... It would be a more consistent design to support this 'feature'.

Reply:
Yep, I think it's a lot better with the change to a map (https://github.com/HSF/prmon/pull/115/commits/466faad930782e5d1745ff259ca7dbd9145a3a61)